File: QA.py
from Enums import AnswerType

class Question():
    def __init__(self, set_param_to, parameter_keyname, question_text, answer_type, answer_options = [], follow_up_question = None) -> None:
        self.question_text = question_text
        self.answer_type = answer_type
        self.answer_options = answer_options
        #self.follow_up_question = follow_up_question
        self.answer = None
        self.set_param_to = set_param_to
        self.parameter_keyname = parameter_keyname
    
    def ask(self):
        # TEMP wait for frontend
        print(self.question_text)
        for a in self.answer_options:
            print(a.text)
        ans = input("> ")

        self.set_answer(ans)
        self.follow_up(int(ans))

    def set_answer(self, ans: str):
        if self.answer_type == AnswerType.MULTI_OPTION:
            self.answer = self.answer_options[int(ans)].value
        elif self.answer_type == AnswerType.NUMERCIAL:
            self.answer = int(ans)
        elif self.answer_type == AnswerType.TEXT:
            self.answer = ans

        self.set_param_to[self.parameter_keyname] = self.answer

    def follow_up(self, ans: int):
        if self.answer_type == AnswerType.MULTI_OPTION:
            follow_up = self.answer_options[ans].follow_up
            if follow_up != None:
                follow_up.ask()
    

class Answer():
    def __init__(self, answer_text, value, follow_up = None) -> None:
        self.text = answer_text
        self.value = value
        self.follow_up = follow_up
        pass

File: codeplate/src/Arduino_Components/Tools/QA.js
import React from "react";
import { AnswerType } from "./Enums.js";

class Question extends React.Component {
    constructor(props) {
        super(props)
        this.state = {
            number: 0,
            selectedValue: null, // Holds the currently selected option
        }
    }
    handleSelection = (value, hasFollowup, followup) => {
        console.log(value)
        this.setState({ selectedValue: value }); // Update local state
        this.props.handleAnswer(value, hasFollowup, followup); // Notify parent if needed
    };
    render() {
        return (
            <div id="question-wrapper">
                <h3>{this.props.questionText}</h3>
                {
                    this.props.answerType === AnswerType.BOOL ?
                        <>
                            <label>
                                <input type= "radio" id = "1" value= "true" onChange={e => this.props.handleAnswer(e.target.value)} />
                                <label for= "boolChoice1"> True </label>
                            </label>
                            <label>
                                <input type= "radio" id = "0" value= "false" onChange={e => this.props.handleAnswer(e.target.value)} />
                                <label form= "boolChoice2"> False </label>
                            </label>
                        </> : ""
                }
                {
                    this.props.answerType === AnswerType.NUMERICAL ?
                        <input type="text" onChange={e => this.props.handleAnswer(e.target.value)} /> : ""
                }
                {
                    this.props.answerType === AnswerType.MULTI_OPTION ?
                    this.props.answerOption.map((option, index) => (
                        <label key={index} className="question-options">
                            <input 
                                type="radio" 
                                id={index}
                                name={this.props.questionText}  // Ensures only one option is selected
                                value={option.value} 
                                checked={this.state.selectedValue === option.value} // Controls selection state
                                onChange={() => this.handleSelection(option.value, option.followup !== "", option.followup || "")} 
                            /> 
                            <label htmlFor={index}>{option.text}</label>
                        </label>
                    )) //TODO: assure only one option can appear as chosen at one time
                    : "" 
                }
                {
                    this.props.answerType === AnswerType.TEXT ?
                        <input type="text" onChange={e => this.props.handleAnswer(e.target.value)} /> : ""
                }
                {
                    this.props.answerType === AnswerType.COLOR ?
                        <input type="color" onChange={e => this.props.handleAnswer(e.target.value)} /> : ""
                }
                <br/>
            </div>
        );
    };
}

const Answer = ({ answerText, value, followUp = null, callbackFunction = null }) => {
    return {
        text: answerText,
        value,
        followUp,
        callbackFunction,
    };
};

export { Question, Answer };